### PR TITLE
sql: add json encoding/decoding support for stmt/txn stats.

### PR DESCRIPTION
**Pull request by Azhng**
Release note: None

### CODE REVIEW

**Comment by Azhng** [comment]
This is the initial commit split off from https://github.com/cockroachdb/cockroach/pull/67090

**Review by jordanlewis — APPROVED** [review]
:lgtm:, @ajwerner did you have any final comments from the last PR?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67805#-:-MfBFbZfEh7OxZCueu4M:b-g35dhh)* status: :shipit: complete! 1 of 0 LGTMs obtained

File: pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_impl.go [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+type jsonDecoder interface {
+	decodeJSON(js json.JSON) error
+}
+
+type jsonEncoder interface {
+	encodeJSON() (json.JSON, error)
+}
+
+type jsonMarshaller interface {
```
Comment (ajwerner) [inline pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_impl.go:32]: nit: the stdlib spells this Marshaler.

**Review by ajwerner — APPROVED** [review]
🚀

**Comment by Azhng** [comment]
TFTR!

bors r+

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67805)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_encoding.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67805#-MfH3a2X2AKP-qO1xwsB:-MfH3a2X2AKP-qO1xwsC:b-plflbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1a4413fc5ff92b9b0faec52acf2d852fb858ff/pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_encoding.go#L57)):*
> ```Go
> //        "distsql":              { "type": "boolean" },
> //        "failed":               { "type": "boolean" },
> //        "opt":                  { "type": "boolean" },
> ```

nit: `opt` is always true and has been for a couple of releases now. Should we remove it everywhere?

**Comment by Azhng** [comment]
@yuzefovich  thanks for pointing it out! I filed this [issue](https://github.com/cockroachdb/cockroach/issues/68077) to track this.

**Pull request merged**